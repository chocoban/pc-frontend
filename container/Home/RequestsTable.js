import { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Table, notification } from 'antd';

import ModalForm from '../Modal/index';
import HomeHeader from './HomeHeader';
import RequestForm from '../Forms/RequestForm';
import {
  fetchRequests, closeModal, postRequest
} from '../../redux/actions';

const columns = [
  {
    title: 'Requests ID',
    dataIndex: 'id',
    sorter: true,
    width: '10%',
    onClick: () => {}
  },
  {
    title: 'Name',
    dataIndex: 'author.name',
    width: '10%',
  },
  {
    title: 'Email',
    dataIndex: 'author.email',
    width: '10%'
  },
  {
    title: 'Country',
    dataIndex: 'author.country',
    filters: [
      { text: 'Uganda', value: 'ug' },
      { text: 'Kenya', value: 'ken' },
      { text: 'Nigeria', value: 'nigeria' }
    ],
    width: '10%',
  }
];

class RequestsTable extends Component {
  state = {
    pagination: {},
    values: {
      amount: '',
      reason: '',
      payment: ''
    }
  };

  handleChange = ({ target }) => {
    this.setState(state => ({
      values: {
        ...state.values,
        [target.name]: target.value
      }
    }
    ));
  };

  isFormValid = (amount, reason, payment) => {
    if (amount !== '' && payment !== '' && reason !== '') {
      return true;
    }
    return false;
  }

  handleSubmit = e => {
    e.preventDefault();
    const {
      amount, reason, payment
    } = this.state.values;
    const {
      closeModal, postRequest
    } = this.props;
    postRequest(amount, reason, payment);
    closeModal();
  };

  renderMissingFieldsNotification = () => {
    notification.open({
      message: 'Missing Fields',
      description: 'Please ensure that all the required fields are filled ',
      duration: 0
    });
  };

  componentDidMount() {
    this.props.fetchRequests();
  }

  componentDidUpdate(prevProps) {
    const { newRequest } = this.props;
    if (prevProps.newRequest !== newRequest && newRequest !== undefined) {
      this.props.fetchRequests();
    }
  }

  render() {
    const {
      amount, reason, payment
    } = this.state.values;
    const {
      visible, closeModal, data
    } = this.props;
    return (
      <Fragment>
        <HomeHeader />
        <ModalForm
          title="Request Form"
          visible={visible}
          handleSubmit={
            this.isFormValid(amount, reason, payment)
              ? this.handleSubmit
              : this.renderMissingFieldsNotification
          }
          handleCloseModal={closeModal}
        >
          <RequestForm
            handleChange={this.handleChange}
          />
        </ModalForm>
        <Table
          columns={columns}
          rowKey={data => data.id}
          dataSource={data}
          pagination={this.state.pagination}
          loading={this.state.loading}
          onChange={this.handleTableChange}
        />
      </Fragment>
    );
  }
}

const mapStateToProps = ({
  modalReducer, requestReducer
}) => ({
  ...modalReducer,
  ...requestReducer
});

const actions = {
  fetchRequests,
  postRequest,
  closeModal
};

export default connect(mapStateToProps, actions)(RequestsTable);
