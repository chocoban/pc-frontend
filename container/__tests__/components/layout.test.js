import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Layout } from 'antd';

import PageLayout from '../../Layout';
import RequestsTable from '../../Home/RequestsTable';

const { Footer } = Layout;

const initialState = {};
const mockStore = configureStore();
const store = mockStore(initialState);

describe('With Enzyme', () => {
  let pageLayoutWrapper;
  beforeAll(() => {
    pageLayoutWrapper = mount(
      <Provider store={ store }>
        <PageLayout />
      </Provider>
    );
  });

  it('PageLayout component snapshot!"', () => {
    expect(pageLayoutWrapper).toMatchSnapshot();
  });

  it('should display requests table component within the Layout', () => {
    const requestTableWrapper = pageLayoutWrapper.find(RequestsTable);
    expect(requestTableWrapper).toBeTruthy();
  });

  it('should display footer component within the Layout', () => {
    const footerWrapper = pageLayoutWrapper.find(Footer);
    expect(footerWrapper.text()).toContain('pettyCash');
  });
});
