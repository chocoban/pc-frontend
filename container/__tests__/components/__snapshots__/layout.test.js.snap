// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Enzyme PageLayout component snapshot!" 1`] = `ReactWrapper {}`;
